VOKMockUrlProtocol.m

```diff
@@ -66,132 +75,148 @@ - (void)startLoading
         || [kHTTPMethodPut isEqualToString:self.request.HTTPMethod]) {
-        NSData *bodyData = self.request.HTTPBody;
-        if (!bodyData && self.request.HTTPBodyStream) {
-            NSMutableData *mutableData = [NSMutableData data];
-            NSInputStream *bodyStream = self.request.HTTPBodyStream;
-            [bodyStream open];
-            NSUInteger length = 0;
-            static NSUInteger const BufferSize = 1024;
-            uint8_t buffer[BufferSize];
-            // Read in chunks of up to BufferSize until there's nothing left to read.
-            do {
-                [mutableData appendBytes:buffer length:length];
-                length = [bodyStream read:buffer maxLength:BufferSize];
-            } while (length > 0);
-            
-            bodyData = mutableData;
-        }
+        NSData *bodyData = [self bodyDataFromRequest:self.request];
         
-        NSString *bodyString = [[NSString alloc] initWithData:bodyData encoding:NSUTF8StringEncoding];
+        // Compute the SHA-256 of the body and generate a resource name from that.
+        NSMutableString *hashResourceName = [resourceName mutableCopy];
+        [hashResourceName appendFormat:AppendSeparatorFormat, [self sha256HexOfData:bodyData]];
         
-        if (bodyString) {
-            NSString *possiblyJSONified = [self alphabeticallyOrderedParametersForJSONString:bodyString];
+        NSString *contentType = [self.request valueForHTTPHeaderField:HTTPHeaderContentType];
+        
+        if ([HTTPHeaderContentTypeFormUrlencoded isEqualToString:contentType]) {
+            // If it's form-URL-encoded, generate a resource name by appending the body as a string.
+            NSString *bodyString = [[NSString alloc] initWithData:bodyData encoding:NSUTF8StringEncoding];
+            [resourceName appendFormat:AppendSeparatorFormat, bodyString];
+            resourceNames = @[ resourceName, hashResourceName, ];
+            
+        } else if ([HTTPHeaderContentTypeJson isEqualToString:contentType]) {
+            // Otherwise, if it's JSON, generate a resource name by bencoding the JSON datastructure and percent-
+            // escaping the resulting string.
+            NSData *bencoded = [VOKBenkode encode:
+                                [NSJSONSerialization JSONObjectWithData:bodyData
+                                                                options:NSJSONReadingAllowFragments
+                                                                  error:NULL]];
+            if (bencoded) {
+                [resourceName appendFormat:AppendSeparatorFormat,
+                 [[[NSString alloc] initWithData:bencoded encoding:NSUTF8StringEncoding]
+                  stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding]];
+                resourceNames = @[ resourceName, hashResourceName, ];
+            } else {
+                // If the bencode fails (or the JSON-decode fails), just use the hash resource name.
```

Comment:
The JSON-decoding of the string could fail if the JSON is malformed.  The bencode could fail for various reasons, such as if the object passed to it contains objects of types that are not bencodable.  One could construct malformed requests that would reliably produce at least the first kind of those failures—I'm not as sure about reliably reproducing the second kind.